src/pudl/transform/ferc1.py

```diff
@@ -5877,133 +5945,61 @@ def add_dimension_total_calculations(
-        # sub-components by broadcast merging the not-total records onto the
-        # new total records.
-        total_w_subdim = (
+    child_candidates = table_dimensions[
+        ~(table_dimensions[dimensions] == "total").any(axis="columns")
+    ]
+
+    total_comps = []
+
+    # check *every* combination of dimensions that could have any total values
+    dim_combos = itertools.chain.from_iterable(
+        itertools.combinations(dimensions, i + 1) for i in range(len(dimensions))
+    )
+    for _total_dims in dim_combos:
+        total_dims = list(_total_dims)
+        parents = meta_w_dims.dropna(subset=total_dims).loc[
+            (meta_w_dims[total_dims] == "total").all(axis="columns")
+        ]
+        if parents.empty:
+            continue
+
+        # There's no wildcard merge in Pandas, so we just ignore whichever
+        # columns have "total"
+        non_total_cols = ["table_name", "xbrl_factoid"] + [
+            d for d in dimensions if d not in total_dims
+        ]
+        total_comps.append(
             pd.merge(
-                # the total records will become _parent columns in new records
-                left=meta_w_dims.loc[(meta_w_dims[dim] == "total")],
-                # the non-total sub-components will become their calculation components
-                right=(table_dimensions[(table_dimensions[dim] != "total")]),
-                on=["table_name", "xbrl_factoid"],
+                left=parents,
+                right=child_candidates,
+                on=non_total_cols,
                 how="inner",
-                validate="m:m",
                 suffixes=("_parent", ""),
             )
-            # The new total -> sub-dimension records must always have the same
-            # parent and component fact/table so overwrite the original values
-            .assign(
-                table_name_parent=lambda x: x.table_name,
-                xbrl_factoid_parent=lambda x: x.xbrl_factoid,
-                weight=1,
-            )
         )
-        subtotal_dfs.append(
-            total_w_subdim.assign(**{f"is_within_dimension_{dim}": True})
-        )
-    # now we have a bunch of *new* records linking total records to their non-total
-    # sub-components.
-    calc_components = pd.concat([calc_components] + subtotal_dfs).reset_index(drop=True)
-    # need to determine if a factoid in a total has multiple totals in its dimensions
-    # bc if so, we are not going to merge on those totals in the table dims table. that
-    # way we will just
-    for dim in dimensions:
-        calc_components = add_has_totals_column_for_dim(
-            calc_components,
-            idx=["table_name_parent", "xbrl_factoid_parent"],
-            dim=f"{dim}_parent",
-        )
-    calc_components = calc_components.assign(
-        total_count_parent=lambda x: x.filter(like="_has_total").sum(axis=1)
+
+    calc_components = pd.concat([calc_components] + total_comps).reset_index(drop=True)
```

Comment:
TODO: restructure code to make clear that we don't actually *use* `calc_components`, we are just smacking a bunch of additional stuff on at the end